MM-54173: Part 2 - unify at mention components

StreakInTheSky opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

A continuation to PR #24402. Like the tickets mentions, wanted "to reuse whatever function normal at mentions is using to create the linking to handle group mentions." So I merged the two components for user at mentions and group mentions. A bit trickier than originally anticipated. Here are the major changes.

- Changed the original class component to the functional component used for the Groups component.
- Factored out a couple utility functions and added tests for them.
   -  `approxGroupPopOverHeight` gets the approximate height of the groups popover.
   -  `getUserOrGroupFromMentionName` figures out whether the mention should be for a user and a group . There was originally another function with that same name that I renamed to `getMentionDetails`, which I thought made more sense. Added tests to both.

Most of the logic for the functions were there, just took them out as standalone functions.

Also because of the change to a functional components I ended up removing some Enzyme tests. They tested for the `placement` state of the popover. Tried to re-implement the tests, but that proved to be too difficult. I came to the conclusion that they weren't actually necessary because they were just testing the component's state. That state was actually coming from a utility function, `popOverOverlayPosition`, that was already being tested for the correct output.

Also misc small changes to clean up the code and handle discriminating unions.

#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/24328
Jira https://mattermost.atlassian.net/browse/MM-54173
-->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
NONE
```

mattermost-build commented:
Hello @StreakInTheSky,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

larkox commented:
We have merged https://github.com/mattermost/mattermost/pull/24091, which adds the "import order" lint rule. This PR touches almost every file under `webapp/channels/src` folder, so you will most surely get a bunch of conflicts. So... a few comments here:
- This PR should only introduce conflicts on the imports area. Anything outside of the imports is something you have to look more closely, since it could be from other PRs.
- My recommendation for solving the conflicts is: accept both changes and delete the duplicates. This allow you to verify both are importing from the same place and that stuff that was moved is still present.
- You will probably need to run `npm run fix` to reorder back the imports after the conflict resolution. Not often, but sometimes you may need to run it twice.

If you have any trouble with the merge, do not hesitate to ask me.

mattermost-build commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

enahum commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,176 @@ type Props = {
-                    </>
-                );
-            }
+    const getPopOver = (user: UserProfile | '', group: Group | '') => {
```
why an empty string? does it makes sense to have it as `undefined` instead?

enahum commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,176 @@ type Props = {
 
-        const suffix = this.props.mentionName.substring(user.username.length);
-        const displayName = displayUsername(user, this.props.teammateNameDisplay);
+        return '';
```
should this be returning null instead ? an empty string is not a component as the two previous returns are

enahum commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,176 @@ type Props = {
-        );
-    }
-}
+                    ) : ''
```
```suggestion
                    ) : null
```

enahum commented on webapp/channels/src/utils/post_utils.ts:
```diff
@@ -723,6 +723,28 @@ export function getUserOrGroupFromMentionName(usersByUsername: Record<string, Us
+    groups: Record<string, Group>,
+    groupsDisabled?: boolean,
+    getMention = getMentionDetails,
+): [UserProfile | '', Group | ''] {
```
I rather be `null` or `undefined` instead of an empty string

StreakInTheSky replied:
Ok. I used an empty string since that's what the original logic was using. Didn't know if it was some convention or something. Would have preferred to use `null`, myself.

StreakInTheSky replied:
ended up using `undefined` so didn't have to use explicit unions everywhere

StreakInTheSky commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,176 @@ type Props = {
     };
 
-    handleClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
+    const handleClick = (group: Group | '') => (e: React.MouseEvent<HTMLAnchorElement>) => {
```
So I'm not actually super familiar with using `useCallback`, but the handlers are used on an `a` tag, not on a child component, would the hook still be useful in that case?

StreakInTheSky commented:
bump

enahum commented on webapp/channels/src/components/at_mention/at_mention.test.tsx:
```diff
@@ -227,52 +226,4 @@ describe('components/AtMention', () => {
         expect(wrapper).toMatchSnapshot();
     });
-
-    test('should have placement state based on ref position of click handler', () => {
```
Why the removal of the placement tests? We should be testing for this to make sure it still opens correctly

StreakInTheSky replied:
The test doesn't work with functional components with React hooks. I tried to see if I could update it for hooks, but was spending way too much time on it without any result. It was really just testing whether the component's `placement` state was updated to the correct string. That string comes from a utility function, which wasn't tested. So instead of banging my head on testing for updated state, created tests for making sure the utility function outputs the correct string.

enahum commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,176 @@ type Props = {
+        return null;
+    };
+
+    const [user, group] = getUserOrGroupFromMentionName(props.mentionName, props.usersByUsername, props.groupsByName, props.disableGroupHighlight);
```
You should probably memoize this

enahum commented:
You also need to fix the check failures

enahum commented:
tests are failing

StreakInTheSky commented:
> tests are failing

ah looks like some snapshot tests with features from master. Merging master and will look into it.

StreakInTheSky commented:
Bump

StreakInTheSky commented:
@calebroseland Hi, looks like this is still waiting for your approval? If there's anything you think still need to be done/fixed, let me know. Thanks!

calebroseland reviewed (changes requested):
Hi @StreakInTheSky, please accept my apologies for my delay in reviewing this.

Overall looks pretty good for combining the two components, just the one request for the the `null` child in `<Overlay>` issue.


(Additionally, thoughts for later)
A few followup changes in `at_mention` (re: preexisting logic from `at_mention_group`) might be good:
1. `target` as a separate piece of state isn't strictly needed if the ref is already there, then `target` wouldn't have to be set or passed around in the functions
2. `const [user, group] = //...` could go at the top of the component—then `group` wouldn't have to be passed around in the functions
3. the dual `Overlay`s seems unnecessary to me, it might be nicer to simplify to just one Overlay

Perhaps something like:
```tsx
//...
{ref.current && (
    <Overlay
        placement={placement}
        show={show}
        target={ref.current}
        rootClose={true}
        onHide={hideOverlay}
        onExited={onHidden}
        shouldUpdatePosition={true}
    >
        {getOverlayContent(user, group, groupUser)}
    </Overlay>
)}
<a
//...
```

StreakInTheSky commented on webapp/channels/src/components/at_mention/at_mention.tsx:
```diff
@@ -32,123 +36,179 @@ type Props = {
-        );
-    }
-}
+                    ) : null
```
oh interesting, wasn't aware of this issue. Noticed the `span` before and thought it wasn't necessary. Will put it back.

StreakInTheSky commented:
So I put the span back and documented why it's there. Also moved the `const [user, group] = ...` part near the top of the component and cleaned up passing `group` through the functions.

Looked into the other things you mentioned and looks like they're all kind of related. Changed it to use only one `Overlay`, but don't have a way to test if it's actually working as intended since it's an enterprise feature and can't really write unit tests to test it. I'll make a new PR for it, so this can be merged?

calebroseland reviewed (approved):
> I'll make a new PR for it, so this can be merged?

Yep, sounds good 👍

calebroseland commented:
/update-branch

lindy65 commented:
Hi @calebroseland - is there a Jira ticket for this PR?  I couldn't find one and the PR referred to at the top of this PR is merged. Not sure whether testing is the same as on the closed Jira ticket?

calebroseland commented:
@lindy65, I don't think there is one, but QA steps based on the code changes for this would be:

Ensure at-mention profile, group, and group-member popovers work in posts:
1. User at-mention profile popover appears
<img width="537" alt="CleanShot 2023-10-12 at 15 39 20@2x" src="https://github.com/mattermost/mattermost/assets/11724372/86f59da3-5fa4-4483-b201-24c9d1070503">

2. Group at-mention popover appears
<img width="541" alt="CleanShot 2023-10-12 at 15 39 25@2x" src="https://github.com/mattermost/mattermost/assets/11724372/dfdf5a5e-0b6f-4693-abbb-06cedd9b5e30">


3. Group at-mention -> group-member popover appears (e.g. after clicking on a group member from group at-mention popover)
<img width="525" alt="CleanShot 2023-10-12 at 15 39 32@2x" src="https://github.com/mattermost/mattermost/assets/11724372/480e1289-a186-4613-a6fb-8d62e36f9caf">

lindy65 commented:
Thanks @calebroseland - I've created https://mattermost.atlassian.net/browse/MM-54932 to track post merge

lindy65 reviewed (approved):
Thanks @StreakInTheSky - looks good to me 👍 

cc @calebroseland

StreakInTheSky commented:
Hi, wondering if this can be merged? Thanks!

lindy65 commented:
> Hi, wondering if this can be merged? Thanks!

cc @enahum @calebroseland
